### PR TITLE
Add "Wind Up" and "Wind Down" mods

### PR DESCRIPTION
**Pull request by LeNitrous**
This implements Wind Up and Wind Down mods available to all modes.
Rate slowly increases or decreases based on map progress and will be at its final rate at 75% of the map progress. Wind Up final rate is based on the Double Time mod and Wind Down final rate is based on the Half Time mod.

Prerequisites (not necessarily required):
- [ ] https://github.com/ppy/osu-framework/issues/2125

Issues:
- [x] New rate exceeds expected final rate.

Preview: https://streamable.com/3qyp6

### CODE REVIEW

File: osu.Game/Rulesets/Mods/ModWindUp.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        public override string Acronym => "WU";
+        public override ModType Type => ModType.Fun;
+        public override FontAwesome Icon => FontAwesome.fa_chevron_circle_up;
+        public override string Description => "Crank it up!";
+        public override double ScoreMultiplier => 1;
+        public override Type[] IncompatibleMods => new[] { typeof(ModDoubleTime), typeof(ModHalfTime) };
+        public abstract double AppendRate { get; }
+    }
+
+    public abstract class ModWindUp<T> : ModWindUp, IUpdatableByPlayfield, IApplicableToClock, IApplicableToBeatmap<T>
```
Comment (peppy) [inline osu.Game/Rulesets/Mods/ModWindUp.cs]: I'd move this to an abstract class called `TimeRampMod`, then have ModWindUp and ModWindDown derive from it.

**Comment by peppy** [comment]
Make sure to address the CI warnings, too: https://ci.appveyor.com/project/peppy/osu/builds/21904357/messages

**Comment by LeNitrous** [comment]
The [`.Update(Playfield playfield)`](https://github.com/ppy/osu/blob/03e57d0423fa25fce086af404a8304b40ec6347a/osu.Game/Rulesets/Mods/ModTimeRamp.cs#L39-L44) method seems to be the reason for both issues.

**Comment by peppy** [comment]
the solution to resetting clocks and fixing the playback control is most likely best addressed with the planned rewrite of the audio subsystem. I've roughly documented the direction we're heading [here](https://github.com/ppy/osu-framework/issues/2125).

**Comment by LeNitrous** [comment]
Made some changes to the mod. It felt awkward that it never truly reached the final rate so I made it 100% of the final rate at 75% of the map's progress.

**Comment by peppy** [comment]
I don't think there's a reason you need per-ruleset versions of these mods. See something like `ModCinema` for reference.

**Comment by LeNitrous** [comment]
I just noticed you no longer need to set the clock rate when you're already adjusting the pitch. Other than that, the pitch should change to its final rate during song select.

**Comment by peppy** [comment]
Possible considerations:

- Do we want to adjust the rate only every 16 beats? ie. is this focused on the skill of tracking an ever-changing bpm or the speed-up in general?
- Do we want to adjust the pitch?

**Comment by LeNitrous** [comment]
- I did some play testing on the mods and timing the hits with a constantly changing bpm proved a bit difficult to keep up with. It made me relied on the approach circles or listening to the audio more which distracts me away from sight reading. This is open for more debate as my plays alone aren't probably enough to justify this point.

- Pitch adjustment is for a more obvious auditory feedback but mostly for aesthetics. Though more opinion regarding this is welcome. For the best of both worlds, I thought of doing what Double Time and Nightcore did and make Inflate and Deflate mods for Wind Up and Wind Down as it's counterparts.

**Comment by peppy** [comment]
Yeah, I agree that the continuous adjust curve definitely calls for a different set of skills than usual, so maybe it's fine to leave it that way.

**Comment by LeNitrous** [comment]
Since mod-specific settings will be a thing (and the fact that there are people who are annoyed by pitch-changed music) I'll amend the variable pitch adjustment for now.

**Review by peppy — APPROVED** [review]

**Pull request merged**